### ISSUE TITLE
Fine-grained iteration count control

### ISSUE THREAD

**Issue by pgera** [google/benchmark#370]
Is there a way to control the iteration count directly instead of --benchmark_min_time ?

**Comment by dominichamon** [comment]
There is not.

An early version of the API had both time controls and iteration count
controls, but the way they worked together was very complicated, and it
turned out that when users were changing iteration counts they almost
always were doing it to ensure a minimum time for the benchmark to run.

What is your use-case?


Dominic Hamon | Google
*There are no bad ideas; only good ideas that go horribly wrong.*

**Comment by pgera** [comment]
The use case is scaling studies where the amount of work done by the benchmark isn't constant. For example, I have some OpenCL benchmarks where I vary the number of work groups and work items, which also affects the amount of actual work the benchmark does (i.e., it is variable work). My argument passing is something like:
b->Args({platform_id, device_id, num_wgs, num_wi_per_wg,region_size});
There are a couple of issues with -benchmark_min_time. One is that it's some combination of CPU time/ wall time, whereas it would be nice to have a manual timer, especially for OpenCL code. The second more important issue is that I want precise control over the number of iterations because a benchmark that takes 100x longer will end up running for 1/100 iterations (in some cases only once).

**Comment by EricWF** [comment]
I agree with @dominichamon that explicit iteration count control is almost always unneeded, and that users are likely to misuse it when using `MinTime` would be appropriate/correct. 

> One is that it's some combination of CPU time/ wall time, whereas it would be nice to have a manual timer,

It's not "some combination", it's CPU time by default unless the user calls `UseRealTime` or `UseManualTime`.

Also the library already supports using user provided/manual timers. Take a look at `UseManualTime` and `SetIterationTime` in `benchmark_api.h`.

**Comment by EricWF** [comment]
That being said I do think there is good rational for adding explicit iteration count control; Specifically to avoid the hidden "timing runs" of the benchmark which are used to determine the number of iterations.

I can imagine writing a benchmark that accesses some global/external state that should only be accessed once, meaning the benchmark may only be run once. Currently this isn't possible since the library will almost always run the benchmark more than once. The only way to avoid this is to allow the user to explicitly provide an iteration count.

For this reason I went ahead and implemented `benchmark::Iterations` in PR #370

**Comment by dominichamon** [comment]
ITYM #373

**Comment by pgera** [comment]
Thanks for adding support for this. This greatly simplifies my workflow.

> Also the library already supports using user provided/manual timers. Take a look at UseManualTime and SetIterationTime in benchmark_api.h

I am using those functions for manual timers, but --benchmark_min_time doesn't tie into the manual timer, right ?

**Comment by EricWF** [comment]
> Edit: Never mind. The manual timer is used. It's just that in my case, the manual time is much smaller than real time, and the iteration count is always determined by (results.real_time_used >= 5 * min_time).

Hmm. That's meant to be a safe guard against malicious benchmarks using little/no CPU time; Preventing them from taking way longer than expected/allowed. I think we should probably disable this condition when a manual timer is used. I'll commit that change tonight.

**Comment by EricWF** [comment]
> I am using those functions for manual timers, but --benchmark_min_time doesn't tie into the manual timer, right ?

`--benchmark_min_time` does tie into the manual timers. It should respect the timer used by the benchmark.

**Comment by EricWF** [comment]
> The manual timer is used. It's just that in my case, the manual time is much smaller than real time, and the iteration count is always determined by (results.real_time_used >= 5 * min_time).

Fixed in master. If the manual timing feature had worked all along would have you needed the explicit iteration count change at all?

**Comment by pgera** [comment]
>Fixed in master. If the manual timing feature had worked all along would have you needed the explicit iteration count change at all?

Thanks! Yes, I think manual iteration count control is still a great feature. I'm not sure if I explained the use case very well. I have a single OpenCL kernel, but I have created several benchmarks out of it based on how the kernel is called. When I run the application with a certain value of min_time, my current output looks like this:

```
Benchmark                                                                                                                                   Time           CPU Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
OclSuite/StridedInterleavedIndependentTest/PlatformId:0/DeviceId:0/NumWGs:1/NumWIperWG:16/NumClsPerWI:256/Stride:1/manual_time          23978 ns     902190 ns       1000
OclSuite/StridedInterleavedIndependentTest/PlatformId:0/DeviceId:0/NumWGs:1/NumWIperWG:16/NumClsPerWI:256/Stride:2/manual_time          23660 ns     864061 ns       1000
...
...
...
OclSuite/StridedInterleavedIndependentTest/PlatformId:0/DeviceId:0/NumWGs:512/NumWIperWG:256/NumClsPerWI:256/Stride:8/manual_time    53616176 ns  170292709 ns          3
OclSuite/StridedInterleavedIndependentTest/PlatformId:0/DeviceId:0/NumWGs:512/NumWIperWG:256/NumClsPerWI:256/Stride:16/manual_time  108993235 ns  344253808 ns          1
```

This would remain similar even with the manual timer's min_time. There is very high difference between the time taken per iteration between the first benchmark and the last one. So it's difficult to pick a good value for min_time, and I would like to ensure that every benchmark is run for at least an acceptable number of iterations. I am also interested in seeing any cache/tlb/other microarchitectural warm-up artifacts. So ideally, I would like to see how the the iteration time varies with number of iterations. Is it possible to get individual iteration's times or aggregate stats ? I know that the documentation mentions --benchmark_repetitions, but that's for inter-benchmark stats and not intra-benchmark stats, right ?

**Comment by EricWF** [comment]
>  it's difficult to pick a good value for min_time, and I would like to ensure that every benchmark is run for at least an acceptable number of iterations. 

You can set different values of `min_time` for different benchmarks using`Benchmark::MinTime(...)`

>  I am also interested in seeing any cache/tlb/other microarchitectural warm-up artifacts. So ideally, I would like to see how the the iteration time varies with number of iterations. Is it possible to get individual iteration's times or aggregate stats ?

If you're trying to expose warm-up artifacts then you likely want to use `--benchmark_repetitions=N` or `Benchmark::Repetitions(...)` instead of specifying a manual iteration count. Each repetitions will run the same amount of iterations, so you can compare the different repetitions by seeing how long they each took.

I'm starting to think you didn't need the ability to explicitly set the iteration count after all, and that @dominichamon was right suggesting that the feature would be a red herring that confuses users out of doing the right thing.

**Comment by EricWF** [comment]
> I know that the documentation mentions --benchmark_repetitions, but that's for inter-benchmark stats and not intra-benchmark stats, right ?

I'm not sure what you mean here, can you clarify?

`--benchmark_repetitions` and `Benchmark::Repetitions` repeat the same benchmark test multiple times with the same number of iterations.

**Comment by pgera** [comment]
>You can set different values of min_time for different benchmarks usingBenchmark::MinTime(...)

That is not very convenient though. I don't know before-hand what the MinTime for each benchmark should be, and even if I did, it is not very feasible to fine tune that per benchmark when I am generating a few thousand benchmarks. It makes much more sense for me to say, "Run each benchmark for a million iterations".

>If you're trying to expose warm-up artifacts then you likely want to use --benchmark_repetitions=N or Benchmark::Repetitions(...) instead of specifying a manual iteration count. Each repetitions will run the same amount of iterations, so you can compare the different repetitions by seeing how long they each took.

I think part of the confusion is arising from the definition of an iteration v/s the definition of a benchmark. Maybe I am using the library wrong, but let's work through this example. Let's say that the benchmark does a random walk in some region of memory (like traversing a link list for instance). In my case, one iteration of this benchmark does one traversal. min_time or num_iterations will dictate the number of times I'll loop over the link list. Let's say that based on min_time that happens to be 10 iterations. The first traversal may be slower than the rest due to cache misses, and I would like to know how much slower that is. If I use --benchmark_repetitions=k, it'll do this exercise k times, but each time it'll run the same 10 iterations, and I won't know what the time delta from iteration 0 to iteration 1 was. The benchmark runs themselves will show relatively little variation. I'm assuming that each benchmark starts with a clean slate. Nothing is carried over between benchmarks, and before each benchmark, hardware data structures are flushed. Do you have any suggestion for getting this data ? You could modify it such that each benchmark only does 1 traversal, and you do --benchmark_repetitions=k, but that is not correct either if you are clearing all state between benchmark runs. You are just getting the timing of the first iteration k times with little to no variation. If you do retain state, you can make it work, but that seems hacky.

Another use case for having precise iteration count control is generating traces (that you later play in a simulator). The most common case is to have just 1 iteration, which can be achieved with min_time=0, but it is not hard to imagine a need for k iterations.

**Comment by EricWF** [comment]
Now that you have the feature do you think you can provide an example that demonstrates the behavior you described?

**Comment by pgera** [comment]
>Now that you have the feature do you think you can provide an example that demonstrates the behavior you described?

Which behaviour in particular would you like to see ?

**Comment by frankist** [comment]
I have a benchmark where I am trying to measure how fast is a reader at dequeuing all elements from a queue. However, the queue has limited size, so in order not to account for dequeue() calls to an empty queue, I have to call dequeue() a limited number of times. I thought of two possible approaches and both seem to require me having control over the number of iterations:
Approach 1 - For each iteration, I could start from a full queue and keep dequeuing until it gets empty, and when that happens move to the next iteration. However, this approach would require me to, before entering the benchmark loop, make several copies of the full queue (each to be used in each iteration). The number of copies would be equal to the number of iterations, which I do not know in advance. I could make that copy inside the loop but this would affect the benchmark timing.
Approach 2 - each iteration corresponds to a single dequeue() call. However, I could not let the number of iterations go beyond the queue size. Not having control over the number of iterations poses a problem once again.

Maybe there is an alternative to these approaches that I am not thinking of. I tried to play with stopTiming() to discount for copies inside the main loop, however, those functions also seem to affect the benchmark timings.

**Comment by dominichamon** [comment]
For approach 1, you can initialize the queue inside the loop while things are paused with a size passed in to the benchmark, and then use the same parameter to dequeue inside the loop. you can also pause the timing while checking for the queue being empty. something like:

```c++
#include <deque>
#include <string>

void BM_dequeue(benchmark::State& state) {
  for (auto _ : state) {
    state.PauseTiming();
    std::deque<std::string> q(state.range(0), "moo");
    state.ResumeTiming();
    while (true) {
      q.pop_back();
      state.PauseTiming();
      if (q.empty()) {
        state.ResumeTiming();
        break;
      }
      state.ResumeTiming();
    }
  }
}
BENCHMARK_REGISTER(BM_dequeue)->Range(100,1000);
```

output:
```
-------------------------------------------------------
Benchmark                Time           CPU Iterations
-------------------------------------------------------
BM_dequeue/8          3644 ns       3645 ns     192098
BM_dequeue/512      203395 ns     204625 ns       3500
BM_dequeue/8192    3188800 ns    3205079 ns        219
BM_dequeue/100       38780 ns      38881 ns      17858
BM_dequeue/512      199370 ns     200510 ns       3517
BM_dequeue/1000     389285 ns     390608 ns       1797
```

while the pausing and resuming does have some small overhead, it is consistent so comparisons between implementations are still going to be useful.

**Comment by frankist** [comment]
"it is consistent so comparisons between implementations are still going to be useful"

That's great! Continuing on that approach, would it be possible in practice to compute its overhead separately and subtract it in the final results? My wild guess would be yes, but computers are weird and a million different variables can affect the speed of your code.

**Comment by dominichamon** [comment]
Measuring the time to stop would then also add to the time, so we'd have to
measure the time it takes to measure...

the overhead should be minimal compared to the operations you're doing. if
not, maybe increase the range to force more work per iteration so that it
is.

Dominic Hamon | Google
*There are no bad ideas; only good ideas that go horribly wrong.*

**Comment by frankist** [comment]
Yes, you are right. The last question was just a personal curiosity.

One more question, I think that if the code between "state.PauseTiming()" and "state.ResumeTiming()" (non-measured) takes much more time than the code I am measuring, google benchmark becomes very slow. My guess is that google benchmark is using the "measured" part of the iteration to compute the number of iterations. If it is fast, the number of iterations will be high. This is causing the slow non-measured part to be computed way too many times. Would this be the expected behavior or a bug?

**Comment by dominichamon** [comment]
That seems reasonable. We want to make sure the measured part is run enough
to get a good signal out. Obviously this leads to the slow part running
more often.

Dominic Hamon | Google
*There are no bad ideas; only good ideas that go horribly wrong.*
